feat: add array datatype on MaxCompute reader

anantadwi13 opened the pull request:
- override getCatalystType to support array datatype on MaxCompute reader (currently not supporting the writer side)
- remove prebuilt custom-dialect.jar (historical retrieval & batch ingestion job will use CustomDialect from the main jar [caraml-spark-application-with-dependencies.jar] instead of custom-dialect.jar, the other usecase which depends on CustomDialect still can download the prebuilt CustomDialect from the release page)
- upgrade com.aliyun.odps:odps-jdbc version to 3.10.1

anantadwi13 committed:
feat: add array datatype on MaxCompute reader
- override getCatalystType to support array datatype on MaxCompute reader (currently not supporting the writer side)
- remove prebuilt custom-dialect.jar (historical retrieval & batch ingestion job will use CustomDialect from the main jar [caraml-spark-application-with-dependencies.jar] instead of custom-dialect.jar, the other usecase which depends on CustomDialect still can download the prebuilt CustomDialect from the release page)
- upgrade com.aliyun.odps:odps-jdbc version to 3.10.0
- upgrade antlr4-runtime version to 4.11.1

anantadwi13 committed:
fix: scala format

shydefoo commented on caraml-store-spark/docker/Dockerfile:
```diff
@@ -2,10 +2,11 @@ FROM --platform=linux/amd64 apache/spark-py:v3.1.3
 ARG HADOOP_ALIYUN_VERSION=3.2.0
 ARG ALIYUN_SDK_OSS_VERSION=2.8.3
 ARG JDOM_VERSION=1.1
+ARG ANTLR4_VERSION=4.11.1
```
What is the reason for upgrading this to 4.11.1?

anantadwi13 replied:
no longer needed now. last time i was upgrading the version to 4.11.1 because the new odps jdbc (v3.10.0) needs that antlr version. but when upgrading the version to 4.11.1 there was an issue with historical retrieval job since it uses python script with pyspark library that depends on older antlr version.

now the odps jdbc v3.10.1 includes the antlr dependency. i check both jobs (historical & batch ingestion) and both are working as expected now

anantadwi13 committed:
fix: resolve antlr dependency issue
- resolving antlr dependency issue by upgrading ali odps jdbc to the patch version
- update datatype validation on historical retrieval script

anantadwi13 committed:
fix: add jackson-databind to fix unit test error

vinoth-gojek commented on caraml-store-pyspark/scripts/historical_feature_retrieval_job.py:
```diff
@@ -463,6 +440,43 @@ def entity_selections(self):
+        }[field.type.lower()]
+    else:
+        return {
+            "bytes": "binary",
```
Curious, what will be a non maxcompute source? I remember bigquery is another one.

anantadwi13 replied:
for non-maxcompute source, It will follow the same behavior as before

vinoth-gojek commented:
Overall, lgtm
